File: src/year2015/day03/part2.cc
#include <set>
#include <string>

#include "../solutions.h"
#include "common.h"
#include "parse.h"

auto Year2015::Day03::Part2(const std::string& input) -> std::string {
  const auto directions = parse(input);
  Coordinate santaPos = {.x = 0, .y = 0};
  Coordinate roboSantaPos = {.x = 0, .y = 0};
  std::set<Coordinate> houses = {santaPos};
  // Start with robo santa as we are changing the current santa before looking
  // at the first instruction
  Coordinate* currentSantaPos = &roboSantaPos;
  for (const auto& direction : directions) {
    if (currentSantaPos == &santaPos) {
      currentSantaPos = &roboSantaPos;
    } else /* if (currentSantaPos == &roboSantaPos) */ {
      currentSantaPos = &santaPos;
    }
    switch (direction) {
      case Direction::North: {
        currentSantaPos->y++;
        houses.insert(*currentSantaPos);
        break;
      }
      case Direction::South: {
        currentSantaPos->y--;
        houses.insert(*currentSantaPos);
        break;
      }
      case Direction::East: {
        currentSantaPos->x++;
        houses.insert(*currentSantaPos);
        break;
      }
      case Direction::West: {
        currentSantaPos->x--;
        houses.insert(*currentSantaPos);
        break;
      }
    }
  }
  return std::to_string(houses.size());
}

File: src/year2015/day03/part2.rs
use std::cell::RefCell;
use std::collections::HashSet;
use std::rc::Rc;

use super::common::Coordinate;
use super::parse::{Direction, parse};

pub fn part2(input: String) -> Result<String, String> {
  let directions = parse(input);
  if directions.is_err() {
    return Err(directions.unwrap_err());
  }
  let directions = directions.unwrap();
  let santa_pos = Rc::new(RefCell::new(Coordinate { x: 0, y: 0 }));
  let robot_pos = Rc::new(RefCell::new(Coordinate { x: 0, y: 0 }));
  let mut curr_pos = Rc::clone(&robot_pos);
  let mut positions: HashSet<Coordinate> = HashSet::new();
  positions.insert(*santa_pos.borrow());

  for direction in directions {
    if Rc::as_ptr(&curr_pos) == Rc::as_ptr(&robot_pos) {
      curr_pos = Rc::clone(&santa_pos);
    } else {
      curr_pos = Rc::clone(&robot_pos);
    }
    match direction {
      Direction::North => curr_pos.borrow_mut().y += 1,
      Direction::South => curr_pos.borrow_mut().y -= 1,
      Direction::East => curr_pos.borrow_mut().x += 1,
      Direction::West => curr_pos.borrow_mut().x -= 1,
    }
    positions.insert(*curr_pos.borrow());
  }
  return Ok(positions.len().to_string());
}
